tools/ci-build/runtime-versioner/src/tag.rs

```diff
@@ -0,0 +1,131 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::{command::CommandExt, repo::Repo};
+use anyhow::{anyhow, bail, Context, Result};
+use smithy_rs_tool_common::release_tag::ReleaseTag;
+use std::str::FromStr;
+use tracing::warn;
+
+/// Discovers and returns the tag of the previous release for the revision at HEAD.
+pub fn previous_release_tag(
+    repo: &Repo,
+    release_tags: &[ReleaseTag],
+    hint: Option<&str>,
+) -> Result<ReleaseTag> {
+    let ancestor_tag = ancestor_tag(repo)?;
+    let hint = hint
+        .map(ReleaseTag::from_str)
+        .transpose()
+        .context("invalid release tag given")?;
+    if let Some(hint) = hint {
+        if !release_tags.contains(&hint) {
+            bail!("specified tag '{hint}' doesn't exist");
+        }
+        if !tag_is_ancestor(repo, &hint)? {
+            bail!("specified tag '{hint}' is not an ancestor to HEAD");
+        }
+        if hint != ancestor_tag {
+            warn!(
+                "expected previous release to be '{ancestor_tag}', \
+                 but '{hint}' was specified. Proceeding with '{hint}'.",
+            );
+        }
+        Ok(hint)
+    } else {
+        Ok(ancestor_tag)
+    }
+}
+
+fn ancestor_tag(repo: &Repo) -> Result<ReleaseTag> {
+    let tag = repo
+        .git(["describe", "--tags"])
+        .expect_success_output("find the current ancestor release tag")?;
+    let maybe_release_tag = ReleaseTag::from_str(&tag);
+    let release_tag = match maybe_release_tag {
+        Ok(tag) => Some(tag),
+        Err(_) => strip_describe_tags_suffix(&tag)
+            .map(ReleaseTag::from_str)
+            .transpose()
+            .context("failed to find ancestor release tag")?,
+    };
+    release_tag.ok_or_else(|| anyhow!("failed to find ancestor release tag"))
+}
+
+// `git describe --tags` appends a suffix if the current commit is not the tagged commit
+//
+// Function assumes the given tag is known to be suffixed.
+fn strip_describe_tags_suffix(tag: &str) -> Option<&str> {
+    // Example release tag with suffix: release-2023-12-01-42-g885048e40
+    let end_index = tag
+        .char_indices()
+        .rev()
+        .skip_while(|(_, c)| *c != '-')
+        .skip(1)
+        .find(|(_, c)| *c == '-')
+        .map(|(i, _)| i);
+    if let Some(end_index) = end_index {
+        Some(&tag[0..end_index])
+    } else {
+        None
+    }
+}
+
+/// Returns all release tags for the repo in descending order by time.
+pub fn release_tags(repo: &Repo) -> Result<Vec<ReleaseTag>> {
+    let mut tags: Vec<_> = repo
+        .git(["tag"])
+        .expect_success_output("find the current ancestor release tag")?
+        .lines()
+        .flat_map(|tag| match ReleaseTag::from_str(tag) {
+            Ok(tag) => Some(tag),
+            Err(_) => {
+                if !tag.starts_with("v0.") {
+                    warn!("ignoring tag '{tag}': doesn't look like a release tag");
+                }
```

Comment:
we have old tags that look like this and `release-*` tags I guess?

Reply:
Yeah. The conditional avoids logging a bunch of useless warnings for our older tags.